File: course_4/module_4/examples/papadimitriou.py
import random
import math
from typing import List, Tuple, Optional


class PapadimitriouTwoSAT:
    def __init__(self, num_variables: int, clauses: List[Tuple[int, int]], seed: Optional[int] = None):
        """
        Initializes the Papadimitriou's 2-SAT solver.
        :param num_variables: Number of variables in the 2-SAT problem.
        :param clauses: List of clauses, where each clause is a tuple of two literals.
                        Positive integers represent variables (e.g., 1 means x1 is True),
                        Negative integers represent negated variables (e.g., -1 means x1 is False).
        :param seed: Optional seed for random number generation for reproducibility.
        """
        self.num_variables = num_variables
        self.clauses = clauses
        if seed is not None:
            random.seed(seed)  # Set the random seed for reproducibility.

    def is_satisfiable(self) -> Tuple[bool, List[Optional[bool]]]:
        """
        Determines if the 2-SAT problem is satisfiable using Papadimitriou's algorithm.
        :return: A tuple (is_satisfiable, assignment).
                 is_satisfiable: True if satisfiable, False otherwise.
                 assignment: A satisfying assignment if one exists, else an empty list.
        """
        max_trials = int(round(math.log2(self.num_variables)))  # Number of independent trials (outer loop).
        max_attempts = 2 * self.num_variables ** 2  # Maximum local moves per trial (inner loop).

        for trial in range(max_trials):
            # Generate a random initial assignment for all variables.
            assignment = [random.choice([True, False]) for _ in range(self.num_variables)]

            for _ in range(max_attempts):
                # Check if the current assignment satisfies all clauses.
                unsatisfied_clauses = [clause for clause in self.clauses if not self._is_clause_satisfied(clause, assignment)]

                if not unsatisfied_clauses:  # If no unsatisfied clauses, return the satisfying assignment.
                    return True, assignment

                # Pick a random unsatisfied clause.
                clause = random.choice(unsatisfied_clauses)

                # Randomly pick one of the two variables in the clause to flip.
                variable_to_flip = abs(random.choice(clause)) - 1  # Convert to 0-based index.
                assignment[variable_to_flip] = not assignment[variable_to_flip]  # Flip the variable.

        # If no satisfying assignment is found, return unsatisfiable.
        return False, []

    @staticmethod
    def _is_clause_satisfied(clause: Tuple[int, int], assignment: List[bool]) -> bool:
        """
        Checks if a clause is satisfied under the current assignment.
        :param clause: A tuple of two literals (e.g., (1, -2)).
        :param assignment: The current variable assignment as a list of booleans.
        :return: True if the clause is satisfied, False otherwise.
        """
        u, v = clause  # Extract the two literals.
        # Determine the truth value of each literal based on the current assignment.
        value_u = assignment[abs(u) - 1] if u > 0 else not assignment[abs(u) - 1]
        value_v = assignment[abs(v) - 1] if v > 0 else not assignment[abs(v) - 1]
        # The clause is satisfied if either literal is True.
        return value_u or value_v


# Example usage:
if __name__ == "__main__":
    # Example 2-SAT problem with 4 variables and 4 clauses.
    num_vars = 4
    example_clauses = [
        (1, 2),     # x1 OR x2
        (-1, 3),    # NOT x1 OR x3
        (3, 4),     # x3 OR x4
        (-2, -4)    # NOT x2 OR NOT x4
    ]

    # Create a solver instance and run the algorithm.
    solver = PapadimitriouTwoSAT(num_vars, example_clauses, seed=42)
    satisfiable, assignment = solver.is_satisfiable()

    if satisfiable:
        print("The 2-SAT problem is satisfiable with the assignment:")
        for i, value in enumerate(assignment, start=1):
            print(f"x{i} = {'True' if value else 'False'}")
    else:
        print("The 2-SAT problem is unsatisfiable.")

File: course_4/module_4/examples/papadimitriou.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"time"
)

// PapadimitriouTwoSAT represents the Papadimitriou's 2-SAT solver.
type PapadimitriouTwoSAT struct {
	NumVariables int      // Number of boolean variables in the problem.
	Clauses      [][2]int // List of clauses, each clause is a pair of literals.
}

// NewPapadimitriouTwoSAT initializes a new instance of Papadimitriou's 2-SAT solver.
// numVariables: Number of variables.
// clauses: List of clauses, where positive integers represent variables (e.g., 1 means x1 is true),
//
//	and negative integers represent negated variables (e.g., -1 means x1 is false).
func NewPapadimitriouTwoSAT(numVariables int, clauses [][2]int) *PapadimitriouTwoSAT {
	return &PapadimitriouTwoSAT{
		NumVariables: numVariables,
		Clauses:      clauses,
	}
}

// IsSatisfiable determines if the 2-SAT problem is satisfiable using Papadimitriou's algorithm.
// Returns a boolean indicating satisfiability and a satisfying assignment if one exists.
func (solver *PapadimitriouTwoSAT) IsSatisfiable() (bool, []bool) {
	maxTrials := int(math.Ceil(math.Log2(float64(solver.NumVariables)))) // Number of independent trials (outer loop).
	maxAttempts := 2 * solver.NumVariables * solver.NumVariables         // Maximum local moves per trial (inner loop).

	rand.Seed(time.Now().UnixNano()) // Seed for random number generation.

	for trial := 0; trial < maxTrials; trial++ {
		// Generate a random initial assignment for all variables.
		assignment := make([]bool, solver.NumVariables)
		for i := range assignment {
			assignment[i] = rand.Intn(2) == 1
		}

		for attempt := 0; attempt < maxAttempts; attempt++ {
			// Find all unsatisfied clauses.
			unsatisfiedClauses := solver.getUnsatisfiedClauses(assignment)

			// If no unsatisfied clauses remain, the problem is satisfiable.
			if len(unsatisfiedClauses) == 0 {
				return true, assignment
			}

			// Pick a random unsatisfied clause.
			clause := unsatisfiedClauses[rand.Intn(len(unsatisfiedClauses))]

			// Randomly pick one of the two variables in the clause to flip.
			variableToFlip := abs(clause[rand.Intn(2)]) - 1          // Convert to 0-based index.
			assignment[variableToFlip] = !assignment[variableToFlip] // Flip the variable.
		}
	}

	// If no satisfying assignment is found after all trials, return unsatisfiable.
	return false, nil
}

// getUnsatisfiedClauses identifies clauses that are not satisfied by the current assignment.
func (solver *PapadimitriouTwoSAT) getUnsatisfiedClauses(assignment []bool) [][2]int {
	var unsatisfied [][2]int
	for _, clause := range solver.Clauses {
		if !isClauseSatisfied(clause, assignment) {
			unsatisfied = append(unsatisfied, clause)
		}
	}
	return unsatisfied
}

// isClauseSatisfied checks if a given clause is satisfied under the current assignment.
// clause: A pair of literals (e.g., (1, -2)).
// assignment: Current assignment of variables.
// Returns true if the clause is satisfied, false otherwise.
func isClauseSatisfied(clause [2]int, assignment []bool) bool {
	u, v := clause[0], clause[1]
	valueU := assignment[abs(u)-1]
	if u < 0 {
		valueU = !valueU
	}
	valueV := assignment[abs(v)-1]
	if v < 0 {
		valueV = !valueV
	}
	return valueU || valueV
}

// abs returns the absolute value of an integer.
func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

// Example usage.
func main() {
	// Example 2-SAT problem with 4 variables and 4 clauses.
	numVars := 4
	clauses := [][2]int{
		{1, 2},   // x1 OR x2
		{-1, 3},  // NOT x1 OR x3
		{3, 4},   // x3 OR x4
		{-2, -4}, // NOT x2 OR NOT x4
	}

	// Create a solver instance.
	solver := NewPapadimitriouTwoSAT(numVars, clauses)

	// Run the algorithm to check satisfiability.
	satisfiable, assignment := solver.IsSatisfiable()

	// Print the results.
	if satisfiable {
		fmt.Println("The 2-SAT problem is satisfiable with the assignment:")
		for i, value := range assignment {
			fmt.Printf("x%d = %t\n", i+1, value)
		}
	} else {
		fmt.Println("The 2-SAT problem is unsatisfiable.")
	}
}
